src/OSPSuite.SimModelNative/src/DESolver.cpp

```diff
@@ -341,6 +341,11 @@ namespace SimModelNative
 					// Reset ODE system (we solve a new one)
 					iResultflag = pSolver->ReInit(solverOutputTime, new_initialvalues_vec);
 
+					for (int iEquation = 0; iEquation < m_ODE_NumUnknowns; iEquation++)
+					{
+						m_ODEVariables[iEquation]->ClearJacobians();
+					}
+
```

Comment:
Because building of Jacobian functions is time consuming, there are some possible performance improvements
1) The ClearJacobians() loop should be called only if `switchUpdate = true`. If the ODE system is restarted for another reason - Jacobian will be the same as before.
2) Jacobians will only change if a **parameter** formula is changed by switch (event). Changing the value of a state variable does not influence it. Also if parameter change is "Value1==>Value2" - Jacobian equation will not change. It will change only if parameter either **was** defined by a formula or **becomes** a (new) formula.
3) Clear only the Jacobians of those state variables which really depend on chenged parameters
<br>

(1) can be easily done right now
(2) also not too complicated - can be done now or in another PR, up to you. Next PR prob better
(3) more work. Would do it **only if** after implementing 1) and 2) there is loss of performance.

Additionaly: if Jacobians were changed during the simulation - they must be cleared at the end as well.
(otherwise if we have a batch run, where the simulation is inititalized once and executed several times with different parameter sets - the 2nd and all further runs would start with  a wrong Jac formulas)

The simulation state for multiple runs is resetted here:
https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/blob/41eb2180fdbeec59e8a343d3f60a906628244c9b/src/OSPSuite.SimModelNative/src/Simulation.cpp#L1096-L1105

Reply:
@Yuri05: after implementing 2, I have noticed we might need to add 3. In the project I have as an example, events are changing parameters constantly. So based on flags, the formulas are reset. The fact is that normally one formula is reset and if we clear all the jacobians every time then the solution is actually worse than the current implementation. We definitely need to clear only the jacobians that are involved on the quantities being changed. I am already looking into this but on Wednesday we could discuss maybe more.

Reply:
ok, I just added this PR: https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/pull/138 
which implements the option 2.5 :)
It checks, which state variables are used in either old or new formula of a parameter (s. comment in the PR).
I think it's sufficent for the performance. The most events just switch parameters from 0 to 1 or 1 to 0; very few events change really the formula of a parameter - and if so: they depend on very few state variables.

The full detection of which state variables depend on changed parameters is complicated.
In fact if you look e.g. at the [current implementation of "which state variables are used in the RHS of a state variable"](https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/blob/41eb2180fdbeec59e8a343d3f60a906628244c9b/src/OSPSuite.SimModelNative/src/SimulationTask.cpp#L290), it is done 2 step:
* in the 1st step, all state variables which are used **in any switch** are collected
* in the 2nd step, RHS dependencies of the initial ODE state are determined, then ALL state variables from the step 1 are added to the dependencies of ALL state variables.

Otherwise one would need to analyze switch dependencies, which might end in circular references and other ugly things (due to the fact that not all events fire at the same time, and so on. So you might have Event1 saying "P2=P3" and Event2 saying "P3=P2", which fire (depending on the state of the system) at different time points and this is perfectly fine. But trying to evaluate the dependencies might become a nightmare then)

Reply:
No, will not work like that. 
I think the only thing which can be easily done for now: check if any parameter is switched from const to non-const or from non-const to const.

Reply:
@abdelr What definitely could be improved: don't calculate Jacobians by those state variables which are definitely zero during the whole simulation, even if something is changed by a switch.
Each state variable caches the indices of those state variables which are (potentially) used in the RHS of this variable:
https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/blob/41eb2180fdbeec59e8a343d3f60a906628244c9b/src/OSPSuite.SimModelNative/include/SimModel/Species.h#L31-L35

So both when calculating the Jacobians initially as well as when clearing/recalculating them: only those should be calculated. All others can be set to const zero.